csrc/multidevice/communication.cpp

```diff
@@ -40,56 +42,148 @@ inline void assertBuffersHaveSameSize(
+  if (hasRoot(params.type)) {
+    auto it = std::find(params.team.begin(), params.team.end(), params.root);
     NVF_ERROR(
-        it != params_.team.end(),
+        it != params.team.end(),
         "root (device ",
-        params_.root,
+        params.root,
         ") must be present in the communication's team");
-    // pytorch's process group expects the root to be specified
-    // as an integer between 0 and world_size-1. We choose it to be
-    // the device's relative index within the team
-    root_relative_index_ = std::distance(params_.team.begin(), it);
   }
+  bool is_root = (my_device_index == params.root);
+  switch (params.type) {
+    case CommunicationType::Gather:
+      assertBufferCount(params.src_bufs, 1);
+      assertBufferCount(params.dst_bufs, is_root ? params.team.size() : 0);
+      break;
+    case CommunicationType::Allgather:
+      assertBufferCount(params.src_bufs, 1);
+      assertBufferCount(params.dst_bufs, params.team.size());
+      break;
+    case CommunicationType::Scatter:
+      assertBufferCount(params.dst_bufs, 1);
+      assertBufferCount(params.src_bufs, is_root ? params.team.size() : 0);
+      break;
+    case CommunicationType::Reduce:
+      assertBufferCount(params.src_bufs, 1);
+      assertBufferCount(params.dst_bufs, is_root ? 1 : 0);
+      break;
+    case CommunicationType::Allreduce:
+      assertBufferCount(params.dst_bufs, 1);
+      assertBufferCount(params.src_bufs, 1);
+      break;
+    case CommunicationType::ReduceScatter:
+      assertBufferCount(params.dst_bufs, 1);
+      assertBufferCount(params.src_bufs, params.team.size());
+      break;
+    case CommunicationType::Broadcast:
+      if (is_root) {
+        assertBufferCount(params.src_bufs, 1);
+        NVF_ERROR(
+            params.dst_bufs.size() < 2, "there must be at most 2 buffer(s)");
+      } else {
+        assertBufferCount(params.src_bufs, 0);
+        assertBufferCount(params.dst_bufs, 1);
+      }
+      break;
+    case CommunicationType::SendRecv:
+      NVF_ERROR(
+          params.team.size() == 1 || params.team.size() == 2,
+          "the team size should be 1 or 2");
+      if (is_root) {
+        assertBufferCount(params.src_bufs, 1);
+        assertBufferCount(params.dst_bufs, (params.team.size() == 1) ? 1 : 0);
+      } else {
+        assertBufferCount(params.src_bufs, 0);
+        assertBufferCount(params.dst_bufs, 1);
+      }
+      break;
+  }
+}
+
+// pytorch's process group expects the root to be specified
+// as an integer between 0 and world_size-1. We choose it to be
+// the device's relative index within the team
+DeviceIdxType getRootRelativeIndex(const CommParams& params) {
+  auto it = std::find(params.team.begin(), params.team.end(), params.root);
+  return std::distance(params.team.begin(), it);
 }
 
+} // namespace
+
+Communication::Communication(IrBuilderPasskey passkey, CommParams params)
+    : Expr(passkey), params_(std::move(params)) {}
```

Comment:
FYI, here's how we enforce certain IR nodes only appearing in the Kernel IR:

https://github.com/NVIDIA/Fuser/blob/main/csrc/kernel_ir.cpp#L834-L836

Again I don't think things like this are important ATM.

Reply:
Thanks! I will add it, but not for now since Communications are registered in a normal fusion right now, inside `MultiDeviceExecutor`